haystack/components/preprocessors/markdown_header_splitter.py

```diff
@@ -0,0 +1,405 @@
+
+        return target_dict
+
+    def _process_split_content(self, split_content: str, split_index: int) -> int:
+        """Process the content of a split and return the number of page breaks."""
+        if not isinstance(split_content, str):
+            return 0
+
+        page_breaks = split_content.count(self.page_break_character)
+        if page_breaks > 0:
+            logger.debug(f"Found {page_breaks} page breaks in split {split_index}")
+        return page_breaks
+
+    def _split_by_function(self, doc: Document) -> List[Document]:
+        """Split document using a custom function that returns dictionaries with 'content' and 'meta'."""
+        logger.debug(f"Splitting document with id={doc.id}")
+        splits = self.splitting_function(doc.content)
+        docs = []
+
+        # calculate total pages and set current page
+        total_pages = doc.meta.get("total_pages", 0) or doc.content.count(self.page_break_character) + 1
+        current_page = doc.meta.get("page_number", 1)
+        logger.debug(f"Starting page number: {current_page}, Total pages: {total_pages}")
+
+        # get meta for each split
+        for i, split in enumerate(splits):
+            meta = {}
+            if doc.meta:
+                meta = self._flatten_dict(doc.meta)
+
+            # add standard metadata
+            meta.update({"source_id": doc.id, "split_id": i, "total_pages": total_pages, "page_number": current_page})
+
+            # get page number based on page breaks
+            page_breaks = self._process_split_content(split["content"], i)
+            current_page += page_breaks
+
+            # add split-specific metadata
+            if split.get("meta"):
+                meta.update(self._flatten_dict(split.get("meta")))
+
+            docs.append(Document(content=split["content"], meta=meta))
+
+        logger.debug(f"Split into {len(docs)} documents for id={doc.id}, final page: {current_page}")
+        return docs
+
+
+@component
+class MarkdownHeaderSplitter:
+    """
+    A custom component that splits documents at markdown headers with optional secondary splitting.
+
+    :param infer_header_levels: If True, attempts to infer and rewrite header levels based on content structure.
+        Useful for documents where all headers use the same level. Defaults to False.
+    :param page_break_character: Character used to identify page breaks. Defaults to form feed ("\\f").
+    :param secondary_split: Optional secondary split condition after header splitting.
+        Options are "none", "word", "passage", "period", "line". Defaults to "none".
+    :param split_length: The maximum number of units in each split when using secondary splitting. Defaults to 200.
+    :param split_overlap: The number of overlapping units for each split when using secondary splitting. Defaults to 0.
+    :param split_threshold: The minimum number of units per split when using secondary splitting. Defaults to 0.
```

Comment:
We normally have these under the init method definition. Could you move them there?